### ISSUE TITLE
Is there any way to pass a HttpClient instance?

### ISSUE THREAD

**Issue by rocboronat** [Baseflow/flutter_cached_network_image#649]
Hello! 👋 

As we try to use the same `HttpClient` to keep keepalive connections, is there any way to pass the `CachedNetworkImageProvider` a `HttpClient` to use?

Thanks 🙇‍♂️

**Comment by PascalOtto** [comment]
I really would like to know this, too. I think it is kind of an essential feature

**Comment by renefloor** [comment]
I'll take you around the CacheManager files.

When creating the CacheManager you supply a config file in [the constructor](https://github.com/Baseflow/flutter_cache_manager/blob/develop/flutter_cache_manager/lib/src/cache_manager.dart#L38-L42). The [default IO config file](https://github.com/Baseflow/flutter_cache_manager/blob/develop/flutter_cache_manager/lib/src/config/_config_io.dart#L12-L24) uses a standard HttpFileService if nothing else is provided. That [http fileservice](https://github.com/Baseflow/flutter_cache_manager/blob/develop/flutter_cache_manager/lib/src/web/file_service.dart#L25-L26) only creates 1 http client if no other client is provided. 

So what you can do is create a different CacheManager object that you supply a config with an HttpFileService that has your httpclient. At the moment you cannot configure this per request.

So like this:
```
import 'package:http/http.dart' as http;
final cacheManager = CacheManager(
  Config(
    'testCache',
    fileService: HttpFileService(
      httpClient: http.Client(),
    ),
  ),
);
```

**Comment by escamoteur** [comment]
how does cached_network_image know which CacheManager it should use?

**Comment by lukehutch** [comment]
If you want to use the cronet or cupertino clients, they cannot be used with `runWithClient`, and to use the [suggested solution](https://github.com/dart-lang/http/blob/321362a99eeffbf83e22e5cbaa5d8356c50f73ee/pkgs/http/lib/src/client.dart#L215) of using `Provider<Client>`, you would have to pass in a new `CacheManager` instance every time you use `CacheWithNetworkImage` (because you have to get the `Client` implementation from the `BuildContext`, using [`context.read<Client>`](https://github.com/dart-lang/http/blob/321362a99eeffbf83e22e5cbaa5d8356c50f73ee/pkgs/flutter_http_example/lib/main.dart#L58C1-L58C38)).

My question is: is it inefficient to reinstantiate a new `CacheManager` and a new `HttpFileService` for every new `CachedNetworkImage` request? Or as long as the `'cacheKey'` matches every time, is the cache going to be reused between requests?

**Comment by lukehutch** [comment]
@renefloor said:

> So what you can do is create a different CacheManager object that you supply a config with an HttpFileService that has your httpclient. At the moment you cannot configure this per request.

but `CachedNetworkImage` does take a `cacheManager` parameter, so why can this not be configured per request?

**Comment by escamoteur** [comment]
If you don't want to deal with context, you can use get_it instead of Provider there.
We use cache_network_image and pass the cache_mamager via get_it without problems.

BTW runWithClient works most of the time for us. We only saw that JNI exception once so far.
Am 27. Juni 2024, 23:16 +0100 schrieb Luke Hutchison ***@***.***>:

**Comment by escamoteur** [comment]
Why would you want to do this on every request?
Am 27. Juni 2024, 23:20 +0100 schrieb Luke Hutchison ***@***.***>:

**Comment by lukehutch** [comment]
> Why would you want to do this on every request?

Because see these comments re context: https://github.com/dart-lang/http/issues/1241#issuecomment-2195765525

I don't think you can just rely on getting the `Client` instance once, and then pre-creating a single instance of `cacheManager`. But I don't fully understand the context issue.

**Comment by lukehutch** [comment]
@escamoteur can you please provide full example code of what you are doing?

**Comment by escamoteur** [comment]
Sure, that is exactly what I do, I register one Client and one Cache Manager in get_it and then I use them everywhere.
I will write a blog post about this in the next days
Am 28. Juni 2024, 00:05 +0100 schrieb Luke Hutchison ***@***.***>:

**Comment by HosseinYousefi** [comment]
> We only saw that JNI exception once so far.

Can I see how you're setting up the `runWithClient`?

**Comment by escamoteur** [comment]
Not at my computer at the moment. Do you get that exception all the time?
Am 28. Juni 2024, 00:13 +0100 schrieb Hossein Yousefi ***@***.***>:

**Comment by HosseinYousefi** [comment]
I don't think we should be getting the exception to begin with! I'm maintaining `package:jni` which is a dependency of `package:cronet_http` and I want to know why this happens.

**Comment by escamoteur** [comment]
I just read a comment of one of our users who said since we added cronnet the app doesn't start anymore. He mentioned that he uses an Android without play services. My hunch is that cronnet might not be available on such systems. Will try tomorrow
Am 28. Juni 2024, 00:18 +0100 schrieb Hossein Yousefi ***@***.***>:

**Comment by HosseinYousefi** [comment]
@escamoteur No worries. Btw are you coming tomorrow to the GDE conf in Berlin? I have a talk there, we can meet and talk about it there!

**Comment by escamoteur** [comment]
Unfortunately not, I m on Tenerife over the summer working from here.
Am 28. Juni 2024, 00:21 +0100 schrieb Hossein Yousefi ***@***.***>:

**Comment by lukehutch** [comment]
> I just read a comment of one of our users who said since we added cronnet the app doesn't start anymore.

@escamoteur was that when you were using `runWithClient`, or did you get the report after you switched to using `get_it`?

**Comment by escamoteur** [comment]
ok, I could just reproduce this, it has nothing to do with `runWithClient` if you use an Android without play services cronnet seems not be available

**Comment by lukehutch** [comment]
> ok, I could just reproduce this, it has nothing to do with `runWithClient` if you use an Android without play services cronnet seems not be available

What is the exact failure, and is there a way to detect that this is going to fail? I want to make my app robust to this situation...

Also, why on earth would cronet require Google Play Services? It is just the Chromium HTTP client library... @HosseinYousefi

**Comment by escamoteur** [comment]
check https://github.com/dart-lang/http/issues/1241#issuecomment-2196475295

I agreee, I too I'm surprised that it seems that cronnet is part of the Google play service part of Android
